app/services/fact_parser.rb

```diff
@@ -80,9 +80,13 @@ def interfaces
 
   # tries to detect primary interface among interfaces using host name
   def suggested_primary_interface(host)
-    # we search among interface with ip and mac if we didn't find it by name
+    # facter 3.x: find 'primary' fact in 'networking' structure
```

Comment:
I think this should live in `app/services/puppet_fact_parser.rb`:
```ruby
def suggested_primary_interface(host)
  facter3_primary = facts.try(:fetch, "networking", nil).try(:fetch, "primary", nil)
  return interfaces[facter3_primary] if facter3_primary
  super
end
```

Reply:
I will move it, however this whole file is completely designed for puppet anyway. It will also break discovery.

Reply:
> I will move it, however this whole file is completely designed for puppet anyway.

If you're talking about `fact_parser.rb` it's quite the opposite. It's an abstract base that can be extended. For Puppet there's `puppet_fact_parser.rb` but for Ansible there's https://github.com/theforeman/foreman_ansible/blob/master/app/services/foreman_ansible/fact_parser.rb.

Reply:
Yeah but `fact_parser.rb` is not abstract at all, it uses Puppet-only facts like ipaddress and others. Abstract class is supposed to be independent of its implementations. But again, I am not going to attack this here.

Reply:
I will agree that it's a weak abstraction. I'd like to have a more neutral implementation. Rather than passing IPs and netmasks I'd love to pass `IPAddr` objects which has all the same information. Each interface would be a hash with `name`, the mac address, the mtu and a list of `IPAddr` objects. Also the type (vlan, bond, ...) if available.

However, we both agree that it's out of scope here.

Reply:
Let's fix reports first and take a look on this as well later on, thanks ;-)